### ISSUE TITLE
Different training behavior between Sequential- and subclassed model (with JAX backend)

### ISSUE THREAD

**Issue by MathiesW** [keras-team/keras#20577]
I wanted to streamline my models by using the Sequential class instead of creating a custom model via subclassing. In my code, I am using a Residual-Wrapper to realize more complex data flows within the Sequential class. Here, I observed exploding gradients when using the Sequential model vs. using the subclassed model with the same model topology, same initialization, and the same data.

I implemented a MWE to look into the issue, and it seems to be a problem with the Keras backend. Here, I create the same model using the Sequential API and the model subclassing. I train both models on the same data and compare the loss. I do this with `jax` and `tensorflow` backend. I don't use `pytorch` so I can not look into it.

**Expected behavior**: The models are identical, the training is deterministic due to global random seed. The loss values should be identical.

**Observed behavior**: With `jax` backend, there is a difference in the loss values between both models. With `tensorflow` backend, both models exhibit exactly the same loss values.

```
import os
os.environ['KERAS_BACKEND'] = 'jax'

from keras import Sequential, Model, layers, initializers
from keras.src.backend.config import backend
from keras import utils
import numpy as np

utils.set_random_seed(1)

conv_kwargs = dict(filters=1, kernel_size=3, padding='same', kernel_initializer=initializers.HeNormal())

class MyModel(Model):
    def __init__(self):
        super().__init__()
        self.layer = layers.Conv1D(**conv_kwargs)

    def build(self, input_shape):
        self.layer.build(input_shape)

    def call(self, inputs):
        return self.layer(inputs)

subclassed_model = MyModel()
sequential_model = Sequential([layers.Conv1D(**conv_kwargs)])

# build models
x = np.ones((1, 32, 1))
subclassed_model.build(x.shape)
sequential_model.build(x.shape)

# train both models on the same data
data = np.random.random((512, 32, 1))  # just some random data to map to itself during training
h = []
for m in [subclassed_model, sequential_model]:
    m.compile(loss='MSE', optimizer='Adam')
    history = m.fit(x=data, y=data, epochs=100, batch_size=32)
    h.append(history)

# get difference between loss values
diff = list(map(lambda item: abs(item[0] - item[1]), zip(h[0].history['loss'], h[1].history['loss'])))

assert not any(diff), f"[{backend()} backend] Deviation between Sequential() and subclassed model"
```

**Comment by james77777778** [comment]
Try this one:

```python
import os

os.environ["KERAS_BACKEND"] = "jax"

import numpy as np

from keras import Model
from keras import Sequential
from keras import backend
from keras import initializers
from keras import layers
from keras import ops

conv_kwargs = dict(
    filters=1,
    kernel_size=3,
    padding="same",
    kernel_initializer=initializers.HeNormal(seed=42),
)


class MyModel(Model):
    def __init__(self):
        super().__init__()
        self.layer = layers.Conv1D(**conv_kwargs)

    def build(self, input_shape):
        self.layer.build(input_shape)

    def call(self, inputs):
        return self.layer(inputs)


subclassed_model = MyModel()
sequential_model = Sequential([layers.Conv1D(**conv_kwargs)])

# build models
subclassed_model.build((1, 32, 1))
sequential_model.build((1, 32, 1))
for a, b in zip(subclassed_model.variables, sequential_model.variables):
    np.testing.assert_allclose(ops.convert_to_numpy(a), ops.convert_to_numpy(b))

# train both models on the same data
data = np.random.random((512, 32, 1))
h = []
for m in [subclassed_model, sequential_model]:
    # JAX backend internally uses `np.random.permutation` to shuffle the dataset
    np.random.seed(42)
    m.compile(loss="MSE", optimizer="Adam")
    history = m.fit(x=data, y=data, epochs=100, batch_size=32)
    h.append(history)

# get difference between loss values
diff = list(
    map(
        lambda item: abs(item[0] - item[1]),
        zip(h[0].history["loss"], h[1].history["loss"]),
    )
)

assert not any(diff), (
    f"[{backend.backend()} backend] Deviation between Sequential() and "
    "subclassed model"
)

print(
    f"[{backend.backend()} backend] No deviation between Sequential() and "
    "subclassed model"
)


```

I got:

```bash
[jax backend] No deviation between Sequential() and subclassed model
```

The root cause is that initializer should be seeded and `np.random.seed` should be called before every `fit` since JAX backend internally uses `np.random.permutation` to shuffle the dataset

**Comment by MathiesW** [comment]
Hi @james77777778, thanks a lot for looking into my code sample. That's an interesting observation. I used `utils.set_random_seed(1)` in my code which I expected to set the seed globally for all keras related stuff. [This example provided by the keras team](https://keras.io/examples/keras_recipes/reproducibility_recipes/) explicitly states that calling `utils.set_random_seed()` sets
- `numpy` random seed
- backend (in my case `jax`)
- python random seed

The weight initialization in my code is deterministic without explicitly setting the seed for the initializer. The shuffling of the dataset obviously not. Is this a bug? I tested on `keras 3.7.0` btw.

**Comment by james77777778** [comment]
I don't think it's a bug. It's expected that multiple calls to `np.random.*` with a single `np.random.seed` will produce different outputs.

A demo here:

```python
import numpy as np

np.random.seed(42)
print(np.random.uniform(size=()))  # 0.3745401188473625
print(np.random.uniform(size=()))  # 0.9507143064099162

np.random.seed(42)
print(np.random.uniform(size=()))  # 0.3745401188473625
```

So, theoretically, your code trains the model without deviation if you train one model at a time. Just make sure not to train both models within the same script.

**Comment by MathiesW** [comment]
Ah i see, that makes total sense. Thanks for the clarification!
